src/input/user_input_comparison.rs

```diff
@@ -311,37 +311,33 @@ where
     }
 
     pub(super) fn comparison_kind(&self, data_id: usize) -> Option<ValueKind> {
-        let Some(kind) = self.comparison_value_kind_buffer.get(&data_id) else {
-            return None;
-        };
-        let Ok(kind) = kind.try_borrow() else {
-            return None;
-        };
-        let Some(kind) = &*kind else {
-            return None;
-        };
-        let kind = kind.iter().map(Clone::clone).collect::<Vec<String>>();
-        let Some(kind) = kind.first() else {
-            return None;
-        };
-        ValueKind::from_str(kind).ok()
+        self.comparison_value_kind_buffer
+            .get(&data_id)?
+            .try_borrow()
+            .ok()?
+            .as_ref()
+            .and_then(|kind| {
+                if let Some(first_kind) = kind.iter().next().cloned() {
+                    ValueKind::from_str(&first_kind).ok()
+                } else {
+                    None
+                }
+            })
     }
 
     pub(super) fn comparison_cmp(&self, data_id: usize) -> Option<ComparisonKind> {
-        let Some(cmp) = self.comparison_value_cmp_buffer.get(&data_id) else {
-            return None;
-        };
-        let Ok(cmp) = cmp.try_borrow() else {
-            return None;
-        };
-        let Some(cmp) = &*cmp else {
-            return None;
-        };
-        let cmp = cmp.iter().map(Clone::clone).collect::<Vec<String>>();
-        let Some(cmp) = cmp.first() else {
-            return None;
-        };
-        ComparisonKind::from_str(cmp).ok()
+        self.comparison_value_cmp_buffer
+            .get(&data_id)?
+            .try_borrow()
+            .ok()?
+            .as_ref()
+            .and_then(|cmp| {
+                if let Some(cmp) = cmp.iter().next().cloned() {
+                    ComparisonKind::from_str(&cmp).ok()
+                } else {
+                    None
+                }
```

Comment:
Since we're already doing our best, it would be great if this part could change like this
```
                kind.iter()
                    .next()
                    .and_then(|kind| ValueKind::from_str(kind).ok())
```

Reply:
Even if we use `if let` ~ `else`, I think the below would be better in order to avoid using `cloned`
```
                if let Some(first_kind) = kind.iter().next() {
                    ValueKind::from_str(first_kind).ok()
                } else {
                    None
                }
```

Reply:
even better, let's do it.

Reply:
You can choose either ` if let` ~ `else` or `and_then`.

Reply:
I'll use `and_then` for clarity and continuity

Reply:
Done